connector/connect/common/src/main/protobuf/spark/connect/base.proto

```diff
@@ -183,6 +183,60 @@ message ExecutePlanResponse {
   }
 }
 
+// Request to transfer client-local artifacts.
+message AddArtifactsRequest {
+
+  // Definition of an Artifact.
+  message Artifact {
+    // The name of the artifact is expected in the form of a "Relative Path" that is made up of a
+    // sequence of directories and the final file element.
+    // Examples of "Relative Path"s: "jars/test.jar", "classes/xyz.class", "abc.xyz", "a/b/X.jar".
+    // The server is expected to maintain the hierarchy of files as defined by their name. (i.e
+    // The relative path of the file on the server's filesystem will be the same as the name of
+    // the provided artifact)
+    string name = 1;
+    // Raw data.
+    bytes data = 2;
+  }
+
+  // A number of small artifacts batched into a single RPC.
+  message Batch {
+    repeated Artifact artifacts = 1;
+  }
+
+  // The client_id is set by the client to be able to collate streaming responses from
+  // different queries.
+  string client_id = 1;
+
+  // User context
+  UserContext user_context = 2;
+
+  // The payload is either a batch of artifacts or a partial chunk of a large artifact.
+  oneof payload {
+    Batch batch = 3;
```

Comment:
This seems unnecessarily complex. Couldn't we just have `repeated Artifact artifacts` here directly? The client can choose to put multiple, or just one. What value is the `Batch` abstraction providing? Why is there an important distinction between a `Batch` and a single artifact?

Reply:
There is a bit of a funny problem here. On the one hand you have jars, they tend to be large and you want to upload them in one or more chunks. On the other hand there are REPL generated classes, these are tiny but there can be many, in this case it makes sense to batch them to reduce RPC overhead.